### PR TITLE
Versions spinner

### PR DESCRIPTION
**Pull request by huwshimi**
Added a spinner when the versions are loading. New spinner style. Use the new spinner for app loading.

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- load the gui
- check that the initial app load shows a new spinner style
- view a charm details page, check that the new spinner appears
- deploy a charm
- view 'Change versions'
- check that the spinner appears when the versions are loading

File: jujugui/static/gui/src/app/components/spinner/_spinner.scss [hunk 1]
```diff
@@ -1,21 +1,72 @@
...
-    @include rotate(360deg);
+    transform: rotate(360deg);
   }
 }
 
-.spinner_loading {
-    @include animation(spinner 1500ms infinite linear);
-    @include create-border-radius(0.5em);
-    @include create-box-shadow(rgba(0, 0, 51, 0.3) 1.5em 0 0 0, rgba(0, 0, 51, 0.3) 1.1em 1.1em 0 0, rgba(0, 0, 51, 0.3) 0 1.5em 0 0, rgba(0, 0, 51, 0.3) -1.1em 1.1em 0 0, rgba(0, 0, 51, 0.3) -1.5em 0 0 0, rgba(0, 0, 51, 0.3) -1.1em -1.1em 0 0, rgba(0, 0, 51, 0.3) 0 -1.5em 0 0, rgba(0, 0, 51, 0.3) 1.1em -1.1em 0 0);
+.spinner-loading {
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/spinner/_spinner.scss]: Update the style of the spinner to look a bit closer what I think I've seen on some mockups.

File: jujugui/templates/index.html.mako [hunk 2]
```diff
@@ -197,32 +199,6 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
       };
 
       var flags = {}; // Declare an empty set of feature flags.
-      startSpinner = function() {
```
Comment (huwshimi) [inline jujugui/templates/index.html.mako]: Use the CSS animated spinner instead of doing it in JS.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3372/
Test PASSed.

File: jujugui/templates/index.html.mako [hunk 3]
```diff
@@ -197,32 +199,6 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
-          left: 'auto' // Left position relative to parent in px
-        };
-        var document = getDocument();
-        var target = document.getElementById('loading-spinner');
-        // If we are in a testing environment, then actually starting the
-        // spinner is unnecessary and counterproductive.
-        if (target.style !== undefined) {
-          spinner = new Spinner(opts).spin(target);
-        }
-      };
```
Comment (jaycee) [inline jujugui/templates/index.html.mako]: Are we using the Spinner js anywhere else? If not could we kill it in this PR?

**Comment by jaycee** [comment]
:+1: 

QA notes:

I played around with several spinner-using-areas of the GUI. All showed a nice spinner.

Thanks @huwshimi. QA OK.

File: jujugui/templates/index.html.mako (hunk 3 above)
Comment (hatched) [inline jujugui/templates/index.html.mako]: +1 after a discussion in standup we are going to remove the flag soon so feel free to cut the old spinner from existence.

**Comment by hatched** [comment]
This looks great, love the component re-use :+1: QA :ok: After removing all of the old spinner bits

File: jujugui/static/gui/src/app/assets/css/_mixins.scss [hunk 4]
```diff
@@ -262,7 +262,8 @@
     color: $cool-grey;
     overflow-x: hidden;
     overflow-y: auto;
-    .spinner_loading {
```
Comment (kadams54) [inline jujugui/static/gui/src/app/assets/css/_mixins.scss]: I was attempting BEM here, with "spinner" as the element and "loading" as the modifier, using a single underscore as a delimiter per https://en.bem.info/method/naming-convention/#modifier-name. I don't really care if we move away to something else, just wanted to explain the rational behind `spinner_loading`.

File: jujugui/static/gui/src/app/components/spinner/_spinner.scss [hunk 5]
```diff
@@ -1,21 +1,72 @@
 @keyframes spinner {
   0% {
-    @include rotate(0deg);
+    transform: rotate(0deg);
   }
   100% {
-    @include rotate(360deg);
+    transform: rotate(360deg);
```
Comment (kadams54) [inline jujugui/static/gui/src/app/components/spinner/_spinner.scss]: Including the `rotate` mixin gave us the vendor prefixes; is it safe to drop that for a straight transform?

File: jujugui/static/gui/src/app/assets/css/_mixins.scss (hunk 4 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/assets/css/_mixins.scss]: Ah yeah, no worries. FYI we use the double underscore, double hyphen style: https://en.bem.info/method/naming-convention/#alternative-naming-schemes

File: jujugui/static/gui/src/app/components/spinner/_spinner.scss (hunk 5 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/spinner/_spinner.scss]: Yep, we could probably remove these mixins as our supported browsers don't need the vendor prefixes.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3377/
Test PASSed.

**Comment by huwshimi** [comment]
Thanks everyone :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**